src/appellate/appellate.js

```diff
@@ -200,3 +206,179 @@ AppellateDelegate.prototype.handleAttachmentPage = async function () {
+
+  this.recap.getAvailabilityForDocuments([this.docId], this.court, callback);
+};
+
+AppellateDelegate.prototype.onDocumentViewSubmit = function (event) {
+  // Security check to ensure message is from a PACER website.
+  if (!PACER.getCourtFromUrl(event.origin)) {
+    console.warn(
+      'Received message from non PACER origin. This should only ' +
+        'happen when the extension is being abused by a bad actor.'
+    );
+    return;
+  }
+
+  let previousPageHtml = copyPDFDocumentPage();
+  let form = document.getElementById(event.data.id);
+
+  let title = document.querySelectorAll('strong')[1].innerHTML;
+  let dataFromTitle = APPELLATE.parsePdfDataFromTitle(title);
+  if (!dataFromTitle) {
+    form.submit();
+    return;
+  }
+  $('body').css('cursor', 'wait');
+  let query_string = new URLSearchParams(new FormData(form)).toString();
+  httpRequest(
+    form.action,
+    query_string,
+    'application/x-www-form-urlencoded',
+    function (type, ab, xhr) {
+      console.info(`RECAP: Successfully submitted RECAP "View" button form: ${xhr.statusText}`);
+
+      const blob = new Blob([new Uint8Array(ab)], { type: type });
+      // If we got a PDF, we wrap it in a simple HTML page.  This lets us treat
+      // both cases uniformly: either way we have an HTML page with an <iframe>
+      // in it, which is handled by showPdfPage.
+      if (type === 'application/pdf') {
+        // canb and ca9 return PDFs and trigger this code path.
+        let html = APPELLATE.getFullPageIframe(URL.createObjectURL(blob));
+        this.showPdfPage(
+          document.documentElement,
+          html,
+          previousPageHtml,
+          dataFromTitle.doc_num,
+          dataFromTitle.att_num,
+          dataFromTitle.case_num
+        );
+      } else {
+        const reader = new FileReader();
+        reader.onload = function () {
+          let html = reader.result;
+          // check if we have an HTML page which redirects the user to the PDF
+          // this was first display by the Northern District of Georgia
+          // https://github.com/freelawproject/recap/issues/277
+          const redirectResult = Array.from(html.matchAll(/window\.location\s*=\s*["']([^"']+)["'];?/g));
+          if (redirectResult.length > 0) {
+            const url = redirectResult[0][1];
+            html = APPELLATE.getFullPageIframe(url);
+          }
+          this.showPdfPage(
+            document.documentElement,
+            html,
+            previousPageHtml,
+            dataFromTitle.doc_num,
+            dataFromTitle.att_num,
+            dataFromTitle.case_num
+          );
+        }.bind(this);
+        reader.readAsText(blob); // convert blob to HTML text
+      }
+    }.bind(this)
+  );
+};
+
+// Given the HTML for a page with an <iframe> in it, downloads the PDF
+// document in the iframe, displays it in the browser, and also
+// uploads the PDF document to RECAP.
+AppellateDelegate.prototype.showPdfPage = async function (
```

Comment:
A lot of this is repeated too, right? Could this be factored out?

Reply:
You're right! I'm going to move this code into a helper function.